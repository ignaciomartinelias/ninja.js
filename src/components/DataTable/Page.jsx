import React from "react";
import { setCurrentPageNumber } from "../../actions/dataTableActions";
import { useSelector } from "react-redux";
import store from "../../store";

const Page = ({ pageNumber, text }) => {
  const currentPageNumber = useSelector(state => state.dataTableReducer.currentPageNumber);

  const isActivePage = currentPageNumber === pageNumber;
  const renderedPageNumber = text ? text : pageNumber + 1;

  const changeToPageNumber = () => {
    store.dispatch(setCurrentPageNumber(pageNumber));
  };

  return (
    <li className='page-item mr-1'>
      <button className={`page-link ${isActivePage ? "button-outline" : ""}`} onClick={changeToPageNumber}>
        {renderedPageNumber}
      </button>
    </li>
  );
};

export default Page;
