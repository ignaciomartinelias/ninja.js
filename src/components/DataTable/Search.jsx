import React from 'react'

const Search = ({ onSearch }) => {

  return (
    <div className="p-b-1">
      <input
        type="search"
        className="form-control"
        placeholder="Search"
        onChange={onSearch} />
    </div>
  )
}

export default Search
