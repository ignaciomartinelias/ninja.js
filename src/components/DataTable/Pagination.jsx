import React from "react";
import { useSelector } from "react-redux";
import Page from "./Page";

const Pagination = ({numberOfPages}) => {
  const currentPageNumber = useSelector(state => state.dataTableReducer.currentPageNumber);

  const pages = [...Array(numberOfPages)].map((i, pageNumber) => {
    return <Page key={pageNumber} pageNumber={pageNumber} />;
  });
  const prev = <Page pageNumber={currentPageNumber - 1} text={"<"} />;
  const next = <Page pageNumber={currentPageNumber + 1} text={">"} />;

  return (
    numberOfPages > 1 && (
      <ul className='pagination'>
        {currentPageNumber - 1 >= 0 && prev}
        {pages}
        {currentPageNumber + 1 < numberOfPages && next}
      </ul>
    )
  );
};

export default Pagination;
