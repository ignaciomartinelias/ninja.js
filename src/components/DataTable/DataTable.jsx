import React, { useState } from "react";
import { useSelector } from "react-redux";
import { setCurrentPageNumber } from "../../actions/dataTableActions";
import store from "../../store";
import Pagination from "./Pagination";
import Search from "./Search";

const DataTable = ({ rows, rowsPerPage = 40, handleSearch, RowToRender }) => {
  const currentPageNumber = useSelector(state => state.dataTableReducer.currentPageNumber);
  const [text, setText] = useState('');

  const getFilteredRows = () => {
    let rowsFound = rows;
    if (text) {
      rowsFound = handleSearch(rows, text);
      store.dispatch(setCurrentPageNumber(0));
    }
    return rowsFound;
  };

  const rowsInPageNumber = pageNumber => {
    const startIndex = pageNumber * rowsPerPage;
    return [startIndex, startIndex + rowsPerPage];
  };

  const filteredRows = getFilteredRows();
  const numberOfPages = rowsPerPage === 0 ? 0 : Math.ceil(filteredRows.length / rowsPerPage);
  const rowsToRender = filteredRows.slice(...rowsInPageNumber(currentPageNumber));

  return (
    <div>
      {handleSearch && <Search onSearch={e => setText(e.target.value)} />}
      <table>
        <tbody>
          {rowsToRender.map(row => (
            <RowToRender key={row.id} row={row} />
          ))}
        </tbody>
      </table>
      <Pagination currentPageNumber={currentPageNumber} numberOfPages={numberOfPages} />
    </div>
  );
};

export default DataTable;
