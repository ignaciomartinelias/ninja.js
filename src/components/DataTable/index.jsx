import React, { useEffect, useState } from 'react'
import { useSelector } from 'react-redux'
import store from '../../store'
import { setCurrentPageNumber, setTotalNumberOfPages } from '../../actions/dataTableActions'

import Pagination from './Pagination'
import Row from './Row'
import Search from './Search'

const DataTable = ({ rowsPerPage }) => {
    const rows = useSelector(state => state.dataTableReducer.rows)
    const currentPageNumber = useSelector(state => state.dataTableReducer.currentPageNumber)
    const totalNumberOfPages = useSelector(state => state.dataTableReducer.totalNumberOfPages)

    const [filteredRows, setFilteredRows] = useState([])
    const [rowsToRender, setRowsToRender] = useState([])

    const search = event => {
        const text = event.target.value
        let rowsFound = rows
        if (text) {
            rowsFound = rows.filter((row) =>
            (row.name1.toLowerCase().includes(text.toLowerCase()) ||
                row.email.toLowerCase().includes(text.toLowerCase()))
            )
        }
        setFilteredRows(rowsFound)
        store.dispatch(setCurrentPageNumber(0))
    }

    const rowsInPageNumber = pageNumber => {
        const startIndex = pageNumber * rowsPerPage
        return [startIndex, startIndex + rowsPerPage]
    }

    useEffect(() => {
        const numberOfPages = rowsPerPage === 0 ? 0 : Math.ceil(filteredRows.length / rowsPerPage)
        numberOfPages !== totalNumberOfPages && store.dispatch(setTotalNumberOfPages(numberOfPages))
        setRowsToRender(filteredRows.slice(...rowsInPageNumber(currentPageNumber)))
    }, [filteredRows, currentPageNumber])

    useEffect(() => {
        setFilteredRows(rows)
    }, [rows])

    return (
        <div>
            <Search onSearch={search} />
            <table>
                <tbody>
                    {rowsToRender.map(row => <Row key={row.per_id} row={row} />)}
                </tbody>
            </table>
            <Pagination />
        </div>
    )
}

DataTable.defaultProps = {
    rowsPerPage: 40
}

export default DataTable;