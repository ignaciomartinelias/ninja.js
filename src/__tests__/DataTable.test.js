import React from "react";
import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import toJson from "enzyme-to-json";
import DataTable from "../components/DataTable";

const mockStore = configureStore();
const store = mockStore({
  dataTableReducer: {
    currentPageNumber: 0,
    totalNumberOfPages: 2,
    rows: [
      {
        name1: "Mads L. Klausen",
        email: "MadsLKlausen@jourrapide.com",
        edit_path: "http://google.com",
        per_id: 1,
      },
      {
        name1: "Alfred K. Krogh",
        email: "AlfredKKrogh@armyspy.com",
        edit_path: "http://google.com",
        per_id: 2,
      },
      {
        name1: "Silas L. Bertelsen",
        email: "SilasLBertelsen@armyspy.com",
        edit_path: "http://google.com",
        per_id: 3,
      },
      {
        name1: "Mia A. Johnsen",
        email: "MiaAJohnsen@dayrep.com",
        edit_path: "http://google.com",
        per_id: 4,
      },
      {
        name1: "Alfred S. Schou",
        email: "AlfredSSchou@jourrapide.com",
        edit_path: "http://google.com",
        per_id: 5,
      },
      {
        name1: "Mads L. Klausen",
        email: "MadsLKlausen@jourrapide.com",
        edit_path: "http://google.com",
        per_id: 6,
      },
      {
        name1: "Alfred K. Krogh",
        email: "AlfredKKrogh@armyspy.com",
        edit_path: "http://google.com",
        per_id: 7,
      },
      {
        name1: "Silas L. Bertelsen",
        email: "SilasLBertelsen@armyspy.com",
        edit_path: "http://google.com",
        per_id: 8,
      },
      {
        name1: "Mia A. Johnsen",
        email: "MiaAJohnsen@dayrep.com",
        edit_path: "http://google.com",
        per_id: 9,
      },
      {
        name1: "Alfred S. Schou",
        email: "AlfredSSchou@jourrapide.com",
        edit_path: "http://google.com",
        per_id: 10,
      },
    ],
  },
});

it("renders 5 rows", () => {
  const wrapper = mount(
    <Provider store={store}>
      <DataTable locale='da' rowsPerPage={5} />
    </Provider>
  );
  expect(wrapper.find("tr").length).toBe(5);
});

it("filters rows based on input", () => {
  const wrapper = mount(
    <Provider store={store}>
      <DataTable locale='da' rowsPerPage={5} />
    </Provider>
  );

  wrapper.find("input").simulate("change", { target: { value: "k" } });
  expect(wrapper.find("tr").length).toBe(4);
});

it("generates 2 pages and first page is active", () => {
  const wrapper = mount(
    <Provider store={store}>
      <DataTable locale='da' rowsPerPage={5} />
    </Provider>
  );
  const pageButtons = wrapper.find("Pagination Page");
  expect(pageButtons.length).toBe(2);
  expect(pageButtons.at(0).find("button").prop("className")).toContain("button-outline");
});

it("Snapshot DataTable Component", () => {
  const wrapper = mount(
    <Provider store={store}>
      <DataTable locale='da' rowsPerPage={5} />
    </Provider>
  );
  expect(toJson(wrapper)).toMatchSnapshot();
});
