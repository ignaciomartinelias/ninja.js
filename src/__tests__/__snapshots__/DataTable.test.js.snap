// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot DataTable Component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <DataTable
    locale="da"
    rowsPerPage={5}
  >
    <div>
      <Search
        onSearch={[Function]}
      >
        <div
          className="p-b-1"
        >
          <input
            className="form-control"
            onChange={[Function]}
            placeholder="SÃ¸g brugere"
            type="search"
          />
        </div>
      </Search>
      <table>
        <tbody>
          <Row
            key="1"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "MadsLKlausen@jourrapide.com",
                "name1": "Mads L. Klausen",
                "per_id": 1,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Mads L. Klausen
                </a>
                <br />
                <small>
                  MadsLKlausen@jourrapide.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="2"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "AlfredKKrogh@armyspy.com",
                "name1": "Alfred K. Krogh",
                "per_id": 2,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Alfred K. Krogh
                </a>
                <br />
                <small>
                  AlfredKKrogh@armyspy.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="3"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "SilasLBertelsen@armyspy.com",
                "name1": "Silas L. Bertelsen",
                "per_id": 3,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Silas L. Bertelsen
                </a>
                <br />
                <small>
                  SilasLBertelsen@armyspy.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="4"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "MiaAJohnsen@dayrep.com",
                "name1": "Mia A. Johnsen",
                "per_id": 4,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Mia A. Johnsen
                </a>
                <br />
                <small>
                  MiaAJohnsen@dayrep.com
                </small>
              </td>
            </tr>
          </Row>
          <Row
            key="5"
            row={
              Object {
                "edit_path": "http://google.com",
                "email": "AlfredSSchou@jourrapide.com",
                "name1": "Alfred S. Schou",
                "per_id": 5,
              }
            }
          >
            <tr>
              <td>
                <a
                  href="http://google.com"
                >
                  Alfred S. Schou
                </a>
                <br />
                <small>
                  AlfredSSchou@jourrapide.com
                </small>
              </td>
            </tr>
          </Row>
        </tbody>
      </table>
      <Pagination>
        <ul
          className="pagination"
        >
          <Page
            key="0"
            pageNumber={0}
          >
            <li
              className="page-item mr-1"
            >
              <button
                className="page-link button-outline"
                onClick={[Function]}
              >
                1
              </button>
            </li>
          </Page>
          <Page
            key="1"
            pageNumber={1}
          >
            <li
              className="page-item mr-1"
            >
              <button
                className="page-link "
                onClick={[Function]}
              >
                2
              </button>
            </li>
          </Page>
        </ul>
      </Pagination>
    </div>
  </DataTable>
</Provider>
`;
